pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -2009,6 +2009,9 @@ func (og *operationGenerator) expandVolumeDuringMount(volumeToMount VolumeToMoun
 				actualStateOfWorld: actualStateOfWorld,
 			}
 			if og.checkForRecoveryFromExpansion(pvc, volumeToMount) {
+				// if recovery feature is enabled, we can use allocated size from PVC status as new size
+				rsOpts.NewSize = pvc.Status.AllocatedResources[v1.ResourceStorage]
+				resizeOp.pluginResizeOpts = rsOpts
```

Comment:
I must be missing something. What is wrong with using `pv.spec.capacity` for NodeExpand call? That's the last size that was confirmed by the storage backend, after a successful ControllerExpand call. `pvc.status.allocatableResource` may contain size not confirmed by the storage yet.

Reply:
The problem is - `pv.Spec.Capacity` can change from what was observed in desired_state_of_world_populator to when actual volume expansion operation starts.  This is only true for **node-only** expandable volumes though, where following scenario can happen:

1. User expanded the PVC from 10Gi to 100Gi.
2. Controller expands successfully(which is basically a NOOP) and `pv.Spec.Capacity` was set to 100Gi.
3. In Kubelet DSOWP sees `pv.Spec.Capacity` to 100Gi and reconciler starts `NodeExpand` to `100Gi`.
4. Now expansion to 100Gi fails on node with one of the final/infeasible errors. 
5. User reduces size to 20Gi.
6. Controller again expands successfully(NOOP) and sets `pv.Spec.Capacity` to 20Gi.
7. Kubelet retries expansion, but hasn't updated DSOW yet with new `pv.Spec.Capacity`.
8. In actual `nodeExpandVolume` operation inside `operation_generator.go`, PVC is fetched again from API server, where we notice `NodeExpansionPending` state and `allocatedResources` is set to `20Gi`.
9. But if we keep using `NewSize` reported by DSOW, then we will `NodeExpand` volume to `100Gi`, which will fail again. 

This behavior esp. breaks new e2e tests I added for **node-only** expandable volumes.

> pvc.status.allocatableResource may contain size not confirmed by the storage yet.

Well, this is not a problem because kubelet **will not** try to expand the volume unless `resizeStatus` was `NodeExpansionPending`, so it is safe to use value reported by `pvc.Status.AllocatedResource`.

Reply:
Having said that - I admit this looks bit ugly right now as a special case. I want to redesign how we detect if volume expansion is needed on the node and we no longer get `NewSize` from DSOW, but merely a boolean flag that, whether expansion is needed or not.  The new size then always gets determined when actual operation starts and that would simplify this *special* looking code.